File: 프로그래머스/lv1/이하은/데이터 분석.cpp
#include <iostream>
#include <string> 
#include <cstring>
#include <sstream>
#include <algorithm>
#include <vector>
#include <cmath>
#include <stack>
#include <queue>
#include <map>
#include <set>
#include <unordered_set>
#include <unordered_map>
using namespace std;
typedef long long ll;
typedef pair<int, int> pii;

string g_ext;
int valExt;
string sortBy;

struct Data {
    int code;
    int date;
    int maximum;
    int remain;
};

vector<vector<int>> answer;
vector<Data> v;

bool compare(Data& a, Data& b){
    if(sortBy == "code"){
        return a.code < b.code;
    }else if(sortBy == "date"){
        return a.date < b.date;
    }else if(sortBy == "maximum"){
        return a.maximum < b.maximum;
    }else{
        return a.remain < b.remain;
    }
}

void removeIfLagerThanExt(){
    function<bool(const Data&)> condition;
    if(g_ext == "code"){        
        condition = [](const Data& obj) {
            return obj.code >= valExt;
        };
    }else if(g_ext == "date"){
        condition = [](const Data& obj) {
            return obj.date >= valExt;
        };
    }else if(g_ext == "maximum"){
        condition = [](const Data& obj) {
            return obj.maximum >= valExt;
        };
    }else{
        condition = [](const Data& obj) {
            return obj.remain >= valExt;
        };
    }
    
    v.erase(
        remove_if(v.begin(), v.end(), condition), 
        v.end()
    );
}

vector<vector<int>> solution(vector<vector<int>> data, string ext, int val_ext, string sort_by) {
    g_ext = ext;
    valExt = val_ext;
    sortBy = sort_by;
    
    for(auto e: data){
        v.push_back({e[0], e[1], e[2], e[3]});
    }
    
    removeIfLagerThanExt();

    sort(v.begin(), v.end(), compare);
    
    for(auto obj: v){
        answer.push_back({obj.code, obj.date, obj.maximum, obj.remain});
    }
    
    return answer;
}

File: 프로그래머스/lv1/이선영/데이터분석.java
import java.util.*;

class Solution {
    public int[][] solution(int[][] data, String ext, int val_ext, String sort_by) {
        int[][] answer = {};
        System.out.println(ext);
        List<int[]> list = new ArrayList<>();
        
        
        for(int i=0; i<data.length; i++){
            switch(ext){
                case "date":
                    if(data[i][1]<val_ext) list.add(data[i]);
                    break;
                case "code":
                    if(data[i][0]<val_ext) list.add(data[i]);
                    break;
                case "maximum":
                    if(data[i][2]<val_ext) list.add(data[i]);
                    break;
                case "remain":
                    if(data[i][3]<val_ext) list.add(data[i]);
                    break;
                default:
                    System.out.println("default");
                    break;
            }
        }
        
        answer= new int[list.size()][4];
        
        for(int i=0; i<list.size(); i++){
            answer[i] = list.get(i);
        }
        
        switch(sort_by){
            case "date":
                System.out.println("date");
                Arrays.sort(answer, (o1,o2)-> o1[1]-o2[1]);
                break;
            case "code":
                Arrays.sort(answer, (o1,o2)-> o1[0]-o2[0]);
                break;
            case "maximum":
                Arrays.sort(answer, (o1,o2)-> o1[2]-o2[2]);
                break;

            case "remain":
                Arrays.sort(answer, (o1,o2)-> o1[3]-o2[3]);
                break;
            default:
                break;
        }
        
        
        return answer;
    }
}
